modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/service/FriendlyURLDLFileEntryLocalServiceWrapper.java

```diff
@@ -0,0 +1,209 @@
+
+		if (_ffFriendlyURLEntryFileEntryConfiguration.enabled()) {
+			List<FriendlyURLEntry> friendlyURLEntries =
+				_friendlyURLEntryLocalService.getFriendlyURLEntries(
+					dlFileEntry.getGroupId(),
+					_classNameLocalService.getClassNameId(FileEntry.class),
+					dlFileEntry.getFileEntryId());
+
+			for (FriendlyURLEntry friendlyURLEntry : friendlyURLEntries) {
+				_friendlyURLEntryLocalService.deleteFriendlyURLEntry(
+					friendlyURLEntry);
+			}
+		}
+
+		return dlFileEntry;
+	}
+
+	@Override
+	public DLFileEntry updateFileEntry(
+			long userId, long fileEntryId, String sourceFileName,
+			String mimeType, String title, String description, String changeLog,
+			DLVersionNumberIncrease dlVersionNumberIncrease,
+			long fileEntryTypeId, Map<String, DDMFormValues> ddmFormValuesMap,
+			File file, InputStream inputStream, long size, Date expirationDate,
+			Date reviewDate, ServiceContext serviceContext)
+		throws PortalException {
+
+		DLFileEntry dlFileEntry = super.updateFileEntry(
+			userId, fileEntryId, sourceFileName, mimeType, title, description,
+			changeLog, dlVersionNumberIncrease, fileEntryTypeId,
+			ddmFormValuesMap, file, inputStream, size, expirationDate,
+			reviewDate, serviceContext);
+
+		if (_ffFriendlyURLEntryFileEntryConfiguration.enabled()) {
+			_updateFriendlyURL(dlFileEntry, title);
+		}
+
+		return dlFileEntry;
+	}
+
+	@Activate
+	@Modified
+	protected void activate(Map<String, Object> properties) {
+		_ffFriendlyURLEntryFileEntryConfiguration =
+			ConfigurableUtil.createConfigurable(
+				FFFriendlyURLEntryFileEntryConfiguration.class, properties);
+	}
+
+	private void _addFriendlyURLEntry(DLFileEntry dlFileEntry)
+		throws PortalException {
+
+		String urlTitle = _friendlyURLEntryLocalService.getUniqueUrlTitle(
+			dlFileEntry.getGroupId(),
+			_classNameLocalService.getClassNameId(FileEntry.class),
+			dlFileEntry.getFileEntryId(), dlFileEntry.getTitle(),
+			LanguageUtil.getLanguageId(LocaleUtil.getSiteDefault()));
+
+		_friendlyURLEntryLocalService.addFriendlyURLEntry(
+			dlFileEntry.getGroupId(),
+			_classNameLocalService.getClassNameId(FileEntry.class),
+			dlFileEntry.getFileEntryId(), urlTitle,
+			ServiceContextThreadLocal.getServiceContext());
+	}
+
+	private Map<String, String> _getUniqueUrlTitleMap(
+		long groupId, long fileEntryId, String title,
+		Map<String, String> urlTitleMap) {
+
+		Map<String, String> newUrlTitleMap = new HashMap<>();
+
+		for (Map.Entry<String, String> entry : urlTitleMap.entrySet()) {
+			String languageId = entry.getKey();
+
+			String urlTitle = urlTitleMap.get(languageId);
+
+			if (Validator.isNotNull(urlTitle) ||
+				((urlTitle != null) && urlTitle.equals(StringPool.BLANK))) {
+
+				urlTitle = _friendlyURLEntryLocalService.getUniqueUrlTitle(
+					groupId,
+					_classNameLocalService.getClassNameId(FileEntry.class),
+					fileEntryId, title, languageId);
+
+				newUrlTitleMap.put(languageId, urlTitle);
+			}
+		}
+
+		return newUrlTitleMap;
+	}
```

Comment:
I think this logic is unnecessary if we only call this when `dlFileEntry.getTitle() != title`. If we apply that logic, we'd simply need to create a new Map from scracth containing the new urlTitle (derived from the title).

Reply:
yeah, we can use updateFriendlyURLEntryLocalization instead updateFriendlyURLEntry and obtain the uniqueUrlTitle as the addition.

Thanks!